core/safeclient/client.go

```diff
@@ -0,0 +1,930 @@
+
+	sub, err := c.Client.SubscribeFilterLogs(ctx, q, proxyC)
+	if err != nil {
+		c.logger.Error("Failed to subscribe to logs", "err", err)
+		return nil, err
+	}
+	c.logger.Info("Subscribed to logs")
+
+	safeSub := NewSafeSubscription(sub)
+	lastBlock := currentBlock
+
+	resubFilterLogs := func() ([]types.Log, error) {
+		currentBlock, err := c.Client.BlockNumber(ctx)
+		if err != nil {
+			c.logger.Error("Failed to get current block number", "err", err)
+			return nil, err
+		}
+		c.logger.Debug("Got current block number for resub", "block", currentBlock)
+
+		if lastBlock >= currentBlock {
+			return nil, nil
+		}
+
+		missedLogs := make([]types.Log, 0)
+		fromBlock := max(lastBlock, currentBlock-c.blockMaxRange) + 1
+
+		for ; fromBlock < currentBlock; fromBlock += (c.blockChunkSize + 1) {
+			toBlock := min(fromBlock+c.blockChunkSize, currentBlock)
+
+			logs, err := c.Client.FilterLogs(ctx, ethereum.FilterQuery{
+				FromBlock: big.NewInt(int64(fromBlock)),
+				ToBlock:   big.NewInt(int64(toBlock)),
+				Addresses: q.Addresses,
+				Topics:    q.Topics,
+			})
+			if err != nil {
+				c.logger.Error("Failed to get missed logs", "err", err)
+				return nil, err
+			} else {
+				c.logger.Info("Got missed logs on resubscribe", "count", len(logs))
+				missedLogs = append(missedLogs, logs...)
+			}
+		}
+
+		return missedLogs, nil
+	}
+
+	resub := func() error {
+		c.clientLock.RLock()
+		defer c.clientLock.RUnlock()
+
+		newSub, err := c.Client.SubscribeFilterLogs(ctx, q, proxyC)
+		if err != nil {
+			c.logger.Error("Failed to resubscribe to logs", "err", err)
+			return err
+		}
+		c.logger.Info("Resubscribed to logs")
+
+		missedLogs, err := resubFilterLogs()
+		if err != nil {
+			c.logger.Error("Failed to get missed logs", "err", err)
+			newSub.Unsubscribe()
+			return err
+		}
+
+		sub = newSub
+		safeSub.SetUnderlyingSub(sub)
+
+		for _, log := range missedLogs {
+			lastBlock = max(lastBlock, log.BlockNumber)
+			ch <- log
+		}
+
+		return nil
+	}
+
+	lastBlock = currentBlock
+
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+
+		ticker := time.NewTicker(c.resubInterval)
+		defer ticker.Stop()
+
+		reinitC := c.WatchReinit()
+
+		for {
+			select {
+			case <-safeSub.Err():
+				c.logger.Debug("Safe subscription ended")
+				return
+			case success := <-reinitC:
+				reinitC = c.WatchReinit()
+				if success {
+					err := resub()
+					if err = c.handleClientError(err); err != nil {
+						c.logger.Error("Failed to resubscribe to logs", "err", err)
+					}
+				}
+			case log := <-proxyC:
+				// since resub pushes the missed blocks directly to the channel and updates lastBlock, this is ordered
+				if lastBlock > log.BlockNumber {
+					continue
+				}
+
+				lastBlock = log.BlockNumber
+				ch <- log
+			case <-ticker.C:
+				c.logger.Debug("Resub ticker fired")
+				err := resub()
+				if err = c.handleClientError(err); err != nil {
+					c.logger.Error("Failed to resubscribe to logs", "err", err)
```

Comment:
```suggestion
					c.logger.Error("Failed to resubscribe to logs", "err", err)
                                        ticker.Reset(c.resubInterval)  
```

Reply:
If triggerReinit() wasn't called within `handleClientError` we shall try to resub with ticker. Otherwise ticker is unused here.